shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Value.hs

```diff
@@ -0,0 +1,407 @@
+-- Values are compared in a lexicographical way
+-- this is needed to define an instance of Relation
+-- use Map Ord instance
+instance Ord (Value crypto) where
+   compare (Value v1) (Value v2) = compare v1 v2
+
+-- TODO this should return the same result as for Value
+instance Ord (CompactValue crypto) where
+   compare v1 v2 = compare (compactValueToValue v1) (compactValueToValue v2)
+
+instance Semigroup (Value crypto) where
+    {-# INLINABLE (<>) #-}
+    (<>) = addv
+
+instance Monoid (Value crypto) where
+    {-# INLINABLE mempty #-}
+    mempty  = zeroV
+    mappend = (<>)
+
+-- instances for CompactValue
+instance Semigroup (CompactValue crypto) where
+    {-# INLINABLE (<>) #-}
+    (<>) v1 v2 = valueToCompactValue $ addv (compactValueToValue v1) (compactValueToValue v2)
+
+instance Monoid (CompactValue crypto) where
+    {-# INLINABLE mempty #-}
+    mempty  = MixValue zeroV
+    mappend = (<>)
+
+--
+-- instance Group (Value crypto) where
+--     {-# INLINABLE inv #-}
+--     inv = scale Integer (Value crypto) (-1)
+
+-- deriving via (Additive (Value crypto)) instance AdditiveSemigroup (Value crypto)
+-- deriving via (Additive (Value crypto)) instance AdditiveMonoid (Value crypto)
+-- deriving via (Additive (Value crypto)) instance AdditiveGroup (Value crypto)
+
+-- instance Module Integer (Value crypto) where
+--     {-# INLINABLE scale #-}
+--     scale i (Value xs) = Value (fmap (fmap (\i' -> i * i')) xs)
+
+-- Linear Map instance
+
+{- note [Assets]
+
+The 'Value' type represents a collection of amounts of different assets.
+
+We can think of 'Value' as a vector space whose dimensions are
+assets. At the moment there is only a single asset type (Ada), so 'Value'
+contains one-dimensional vectors. When asset-creating transactions are
+implemented, this will change and the definition of 'Value' will change to a
+'Map Asset Int', effectively a vector with infinitely many dimensions whose
+non-zero values are recorded in the map.
+
+To create a value of 'Value', we need to specifiy an asset policy. This can be done
+using 'Ledger.Ada.adaValueOf'. To get the ada dimension of 'Value' we use
+'Ledger.Ada.fromValue'. Plutus contract authors will be able to define modules
+similar to 'Ledger.Ada' for their own assets.
+
+-}
+
+
+{-# INLINABLE valueOf #-}
+-- | Get the quantity of the given currency in the 'Value'.
+valueOf :: Value crypto -> PolicyID crypto -> AssetID -> Integer
+valueOf (Value mp) cur tn =
+    case Data.Map.Strict.lookup cur mp of
+        Nothing -> 0 :: Integer
+        Just i  -> case Data.Map.Strict.lookup tn i of
+            Nothing -> 0
+            Just (Quantity v)  -> v
+
+{-# INLINABLE symbols #-}
+-- | The list of 'PolicyID's of a 'Value'.
+symbols :: Value crypto -> [PolicyID crypto]
+symbols (Value mp) = keys mp
+
+{-# INLINABLE singleType #-}
+-- | Make a 'Value' containing only the given quantity of the given currency.
+singleType :: PolicyID crypto -> AssetID -> Integer -> Value crypto
```

Comment:
Why not `singleton`? In this instance I think the plutus name is good: that's what e.g. `Map` uses for a map with a single entry.

Reply:
Also should take a Quantity.